### PR TITLE
ztest: add test summary support.

### PR DESCRIPTION
**Pull request by smrtos**
This is a second try of https://github.com/zephyrproject-rtos/zephyr/pull/48105.
If this PR is merged, https://github.com/zephyrproject-rtos/zephyr/pull/48484 can be closed.

Several issues are addressed:

- https://github.com/zephyrproject-rtos/zephyr/issues/48480
  **Root cause**: When declaring a unit test with ZTEST* macros, a stats object is defined and associated with the unit test. Previously, this stats object is named only after the test function. So if the same function is used in different suites, the same stats object will be defined multiple times, thus the duplicate symbol error. This can be avoided by including the suite name in the stats object name because suite name is guaranteed to be unique. And also, this can make sure test function in different suites can have separate stats.
  **Fixed with**: https://github.com/smrtos/zephyr/blob/d36d46cdf61408761e297309aaaafd760fe3dbb3/subsys/testsuite/ztest/include/zephyr/ztest_test_new.h#L266
  **Verified with**:
  https://github.com/smrtos/zephyr/tree/ztest_add_summary/tests/ztest/summary

- https://github.com/zephyrproject-rtos/zephyr/pull/48105#pullrequestreview-1055942751
  **Root cause**: should be caused by the messy [log flushing](https://github.com/smrtos/zephyr/blob/d36d46cdf61408761e297309aaaafd760fe3dbb3/subsys/testsuite/ztest/src/ztest_new.c#L863) when log panic is triggered.
  **Fixed with**: https://github.com/smrtos/zephyr/blob/d36d46cdf61408761e297309aaaafd760fe3dbb3/subsys/testsuite/ztest/src/ztest_new.c#L862
   **Verified with**:
   twister -p qemu_x86_64 -s tests/subsys/logging/log_api/logging.log_api_deferred_printk_cpp -vv
   twister -p qemu_riscv32_smp -s tests/subsys/logging/log_api/logging.log_api_deferred_printk_cpp -vv
   twister -p qemu_cortex_a53_smp -s tests/subsys/logging/log_api/logging.log_api_deferred_printk_cpp -vv
  
- https://github.com/zephyrproject-rtos/zephyr/issues/48469
  **Root cause**: the tests/kernel/sched/schedule_api is still using old ztest. Previous PR didn't treat them fair enough...
  **Fixed with**: lines like: https://github.com/smrtos/zephyr/blob/d36d46cdf61408761e297309aaaafd760fe3dbb3/subsys/testsuite/include/zephyr/tc_util.h#L133-L142
   **Verified with**:
   west build -b qemu_cortex_a53  tests/kernel/sched/schedule_api --pristine &&  west build -t run
   twister -p qemu_x86_64 -s tests/kernel/sched/schedule_api/kernel.scheduler -vv
   twister -p qemu_riscv32_smp -s tests/kernel/sched/schedule_api/kernel.scheduler -vv
   twister -p qemu_cortex_a53_smp -s tests/kernel/sched/schedule_api/kernel.scheduler -vv
   These runs will not have summary output because it is using old ztest.

- Also fixed an old bug: https://github.com/zephyrproject-rtos/zephyr/issues/48547

The summary looks like this now:

```
DEBUG   - QEMU (2937061): ------ TESTSUITE SUMMARY START ------
DEBUG   - QEMU (2937061): SUITE PASS - 100.00% [suite1]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.001 seconds
DEBUG   - QEMU (2937061): - PASS - [suite1.test_foo] duration = 0.001 seconds
DEBUG   - QEMU (2937061):
DEBUG   - QEMU (2937061): SUITE SKIP -   0.00% [suite2]: pass = 0, fail = 0, skip = 1, total = 1 duration = 0.001 seconds
DEBUG   - QEMU (2937061): - SKIP - [suite2.test_foo] duration = 0.001 seconds
DEBUG   - QEMU (2937061):
DEBUG   - QEMU (2937061): ------ TESTSUITE SUMMARY END ------
```
And: 
```
DEBUG   - QEMU (2936214): ------ TESTSUITE SUMMARY START ------
DEBUG   - QEMU (2936214): SUITE PASS - 100.00% [test_log_api]: pass = 8, fail = 0, skip = 1, total = 9 duration = 14.851 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_arg_evaluation] duration = 2.422 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_arguments] duration = 1.210 seconds
DEBUG   - QEMU (2936214): - SKIP - [test_log_api.test_log_backend_runtime_filtering] duration = 1.209 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_from_declared_module] duration = 2.420 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_overflow] duration = 3.630 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_override_level] duration = 1.320 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_panic] duration = 1.212 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_printk] duration = 0.109 seconds
DEBUG   - QEMU (2936214): - PASS - [test_log_api.test_log_various_messages] duration = 1.319 seconds
DEBUG   - QEMU (2936214):
DEBUG   - QEMU (2936214): SUITE SKIP -   0.00% [test_log_api_1cpu]: pass = 0, fail = 0, skip = 1, total = 1 duration = 1.211 seconds
DEBUG   - QEMU (2936214): - SKIP - [test_log_api_1cpu.test_log_msg_dropped_notification] duration = 1.211 seconds
DEBUG   - QEMU (2936214):
DEBUG   - QEMU (2936214): ------ TESTSUITE SUMMARY END ------
```

Signed-off-by: Ming Shao <ming.shao@intel.com>

### CODE REVIEW

**Comment by smrtos** [comment]
Please do not merge this PR for now.

Reasons:
- Need time to test on more platforms with the whole tests.
- Need to add time stats for test execution.

**Comment by smrtos** [comment]
Also fix: https://github.com/zephyrproject-rtos/zephyr/issues/48547

**Comment by smrtos** [comment]
The main changes are done.
Now testing with as many boards as I can find.
The tests coding style are quite versatile.
Hope everything is fine...

**Comment by smrtos** [comment]
The CI saw below error:
https://github.com/zephyrproject-rtos/zephyr/runs/7627365444?check_suite_focus=true


But I tried with:
```
twister -p qemu_arc_hs -s tests/lib/mem_alloc/libraries.libc.newlib.mem_alloc -vv
```
All passed:
```
DEBUG   - QEMU (409867): *** Booting Zephyr OS build zephyr-v3.1.0-2464-g928656470a57  ***
DEBUG   - QEMU (409867): Running TESTSUITE test_c_lib_dynamic_memalloc
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_malloc_align
DEBUG   - QEMU (409867): Compiler type info for arc qemu_arc
DEBUG   - QEMU (409867): TYPE            SIZE  ALIGN
DEBUG   - QEMU (409867): int                4      4
DEBUG   - QEMU (409867): long               4      4
DEBUG   - QEMU (409867): long long          8      4
DEBUG   - QEMU (409867): double             8      4
DEBUG   - QEMU (409867): size_t             4      4
DEBUG   - QEMU (409867): void *             4      4
DEBUG   - QEMU (409867): void (*)(void)     4      4
DEBUG   - QEMU (409867): time_t             8      4
DEBUG   - QEMU (409867): PASS - test_malloc_align in 0.167 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_malloc
DEBUG   - QEMU (409867): PASS - test_malloc in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_free
DEBUG   - QEMU (409867): PASS - test_free in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_calloc
DEBUG   - QEMU (409867): SKIP - test_calloc in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_realloc
DEBUG   - QEMU (409867): PASS - test_realloc in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_reallocarray
DEBUG   - QEMU (409867): SKIP - test_reallocarray in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_memalloc_all
DEBUG   - QEMU (409867): PASS - test_memalloc_all in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): START - test_memalloc_max
DEBUG   - QEMU (409867): PASS - test_memalloc_max in 0.001 seconds
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): TESTSUITE test_c_lib_dynamic_memalloc succeeded
DEBUG   - QEMU (409867): ===================================================================
DEBUG   - QEMU (409867): RunID: be74c1058a36f2326e4b62bb09a9f19a
DEBUG   - QEMU (409867): PROJECT EXECUTION SUCCESSFUL
DEBUG   - QEMU (409867) complete (passed) after 2.20586895942688 seconds

```

**Comment by smrtos** [comment]
For the CI error with [qemu_arc_hs](https://github.com/zephyrproject-rtos/zephyr/runs/7629029614?check_suite_focus=true):

I rebased and tested with below commands:

`twister -p qemu_arc_hs -T tests`

Still all passed. Cannot reproduce the CI error.

```
INFO    - 1327 test scenarios (1234 configurations) selected, 810 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete: 1234/1234  100%  skipped:  914, failed:    0
INFO    - 320 of 1234 test configurations passed (100.00%), 0 failed, 914 skipped with 0 warnings in 1241.45 seconds
INFO    - In total 11058 test cases were executed, 8819 skipped on 1 out of total 478 platforms (0.21%)
INFO    - 292 test configurations executed on platforms, 28 test configurations were only built.
```

For the CI error with [mps2_an385](https://github.com/zephyrproject-rtos/zephyr/runs/7629029255?check_suite_focus=true), I don't have such a board to test.

**Comment by nashif** [comment]
> For the CI error with [mps2_an385](https://github.com/zephyrproject-rtos/zephyr/runs/7629029255?check_suite_focus=true), I don't have such a board to test.

this is qemu

**Comment by smrtos** [comment]
I just re-pushed.  It's strange that the mps2_an385 failure disappeared. 
Only the `qemu_arc_hs` failure remains: https://github.com/zephyrproject-rtos/zephyr/runs/7644514469?check_suite_focus=true

**Comment by smrtos** [comment]
I just re-pushed. No significant change.

It's strange that the `mps2_an385` CI failure disappeared. 

Only the `qemu_arc_hs` failure remains: https://github.com/zephyrproject-rtos/zephyr/runs/7647515326?check_suite_focus=true
Which still all pass on my local.

`twister -vv -p qemu_arc_hs -T tests/`

```
INFO    - 320 of 1234 test configurations passed (100.00%), 0 failed, 914 skipped with 0 warnings in 1214.80 seconds
INFO    - In total 11058 test cases were executed, 8819 skipped on 1 out of total 478 platforms (0.21%)
INFO    - 292 test configurations executed on platforms, 28 test configurations were only built.
```

**Comment by smrtos** [comment]
Seems the CI healed itself... And I cannot find the CI log for qemu_arc_hs...

**Review by nashif — DISMISSED** [review]

**Comment by nashif** [comment]
> Seems the CI healed itself... And I cannot find the CI log for qemu_arc_hs...

no it did not :)
I just triggered a re-run. Some qemu platforms fails occasionally due to timing issues with the host...

**Comment by smrtos** [comment]
Testing on reel_board and found some build failure like below. Investigating.

```
  Manually-specified variables were not used by the project:

    TC_RUNID
```

**Comment by smrtos** [comment]
Saw some failures on reel_board. Related bug: https://github.com/zephyrproject-rtos/zephyr/issues/48725
I don't think it is related to this PR.

**Comment by smrtos** [comment]
Verification so far:

`twister -vv -p qemu_arc_hs -T tests/`   -- 100% pass
`twister -vv -p qemu_cortex_a53_smp -T tests` -- 100% after manual retry
`twister -vv --hardware-map ~/daily_qa/cavs25.map --device-testing -T tests` -- pass with bugs on [main branch](https://github.com/zephyrproject-rtos/zephyr/pull/48499#issuecomment-1206111226).

**Review by jeremybettis — DISMISSED** [review]

File: subsys/testsuite/ztest/Kconfig
Comment (aaronemassey) [inline subsys/testsuite/ztest/Kconfig:136]: nit:

"shwon" should be "shown"

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (aaronemassey) [inline subsys/testsuite/include/zephyr/tc_util.h:130]: These configs are getting a bit deep. Can we add comments at each `endif` like:

`#endif /* CONFIG_ZTEST_NEW_API */`

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (aaronemassey) [inline subsys/testsuite/include/zephyr/tc_util.h:142]: See other comment on endif coments.

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (aaronemassey) [inline subsys/testsuite/include/zephyr/tc_util.h:166]: See other comment on endif coments.

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (aaronemassey) [inline subsys/testsuite/ztest/src/ztest_new.c:694]: Please double check my algebra on this: But I think passrate2 is always equal to zero here.

`passrate` == `passrate1 * (distinct_total - distinct_skip)`

**Comment by aaronemassey** [comment]
Reviewing right now, I'm almost done but please DNM as I have some questions. Sorry for not getting to this earlier.

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (aaronemassey) [inline subsys/testsuite/ztest/src/ztest_new.c:731]: Won't this code just be run once?

`flush_frequency` starts out at `0` and then we increment it to `1` but then never change it so it stays at `1` and this branch only taken once ever since `(1 % 1) == 0` is a false predicate.

**Review by aaronemassey — CHANGES_REQUESTED** [review]

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (smrtos) [reply subsys/testsuite/ztest/src/ztest_new.c:731]: But I do have a `flush_frequency++` at line 734 and it is `%2`.

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (smrtos) [reply subsys/testsuite/ztest/src/ztest_new.c:694]: All `passrate, passrate1, passrate2` are of int type. The line 693 is integer division.
For example: suppose `distinct_pass=1234`, `distinct_total - distinct_skip = 10000`, then `passrate=1234*100=123400`, `passrate1= 123400/10000 = 12`, `passrate2 = (123400 - 12*10000)/100 = 34`.
The final output on line 697 will be ` 12.34%`.

Comment (smrtos) [reply subsys/testsuite/ztest/src/ztest_new.c]: At first I thought this was just integer division and tried with below:
Suppose `distinct_pass=1234`, `distinct_total - distinct_skip = 10000`,
then `passrate=1234*100=123400`, 
`passrate1= 123400/10000 = 12`,
`passrate2 = (123400 - 12*10000)/100 = 34`.
The final output on line 697 will be `  12.34%`.

But it turns out not fully correct for other scenarios. Such as `2/3=66.67%`.
And I didn't consider the rounding in the last implementation.

Thanks for pointing this out! I have refined it. Could you have another review.
The new calculation is like this:
Say:
```
distinct_total=3;
distinct_pass=2;
distinct_skip=distinct_fail=0;
```
Then:
```
essential_total=3-0=3;
passrate = 2*100000 = 200000
passrate0 = 200000/3 = 66666
passrate1 = 66666 / 1000 = 66
passrate2 = (66666 -  66*1000)/10=66
passrate3 = 66666 - 66*1000 - 66*10 = 6
passrate3 >=5, so passrate2++
```
At last we got: `passrate1.passrate2% =  66.67%`

**Comment by smrtos** [comment]
@nashif @jeremybettis  @aaronemassey  Could you help review again? I refined a bit.

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (smrtos) [reply subsys/testsuite/include/zephyr/tc_util.h:130]: done.

**Review by jeremybettis — DISMISSED** [review]

File: subsys/testsuite/ztest/Kconfig
Comment (smrtos) [reply subsys/testsuite/ztest/Kconfig]: Done. Thanks!

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (smrtos) [reply subsys/testsuite/include/zephyr/tc_util.h:142]: done.

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (smrtos) [reply subsys/testsuite/include/zephyr/tc_util.h:166]: Done.

**Comment by aaronemassey** [comment]
> @aaronemassey @nashif @jeremybettis Could you help review again? I refined a bit. Thanks.

Might get to it on the weekend, but if not will take a look Monday!

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (smrtos) [reply subsys/testsuite/ztest/src/ztest_new.c:738]: Yes, you are right. The `flush_frequency++` should be outside of the loop.

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (aaronemassey) [reply subsys/testsuite/include/zephyr/tc_util.h:130]: Apologies, I think I wasn't clear enough in my prior request:

Could we other `#endif /* CONFIG_{SOMETHING}` ? I didn't just mean `CONFIG_ZTEST_NEW_API` but the others as well.

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (aaronemassey) [reply subsys/testsuite/ztest/src/ztest_new.c]: Ahh I understand more now. It'd help if we renamed the passrate variables with more descriptive terms representing their intermediary values though. It'd make it easier for the next person wanting to look at or change this code for whatever reason.

**Review by aaronemassey — CHANGES_REQUESTED** [review]
See conversation comments. Just renaming and adding more `#endif` comments and then I'll approve.

File: subsys/testsuite/ztest/src/ztest_new.c
Comment (smrtos) [reply subsys/testsuite/ztest/src/ztest_new.c]: Indeed, I don't like those variable names either. English is not my native language but I will try. It's like this now:

https://github.com/smrtos/zephyr/blob/becee84476b3606e7912178c0fea459c9ba77330/subsys/testsuite/ztest/src/ztest_new.c#L694-L702

File: subsys/testsuite/include/zephyr/tc_util.h
Comment (smrtos) [reply subsys/testsuite/include/zephyr/tc_util.h:130]: As you wish. :)

Comment (smrtos) [reply subsys/testsuite/include/zephyr/tc_util.h:130]: As you wish. :)

**Review by aaronemassey — DISMISSED** [review]
Awesome! Thank you for this!

**Comment by nashif** [comment]
@smrtos  please rebase

**Comment by smrtos** [comment]
Rebased. 

The ztest_new added the skip handling. So I don't need to do it.

image

**Review by nashif — APPROVED** [review]

**Review by chen-png — APPROVED** [review]

**Pull request merged**
